modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/gossip/GossipDaemon.scala

```diff
@@ -74,25 +77,42 @@ object GossipDaemon {
         case (hash, signedRumor) =>
           signedRumor.value.hashF
             .map(_ === hash)
-            .flatTap { valid =>
-              if (valid) Applicative[F].unit
-              else
-                logger.warn(
+            .flatTap(
+              logger
+                .warn(
                   s"Discarding rumor ${signedRumor.value.show} with hash ${hash.show} due to invalid hash"
                 )
-            }
+                .unlessA
+            )
       }
 
     private def validateSignature(batch: RumorBatch): F[RumorBatch] =
       batch.filterA {
         case (hash, signedRumor) =>
-          signedRumor.hasValidSignature.flatTap { valid =>
-            if (valid) Applicative[F].unit
-            else
-              logger.warn(
-                s"Discarding rumor ${signedRumor.value.show} with hash ${hash.show} due to invalid hash signature"
-              )
-          }
+          signedRumor.hasValidSignature
+            .flatTap(
+              logger
+                .warn(
+                  s"Discarding rumor ${signedRumor.value.show} with hash ${hash.show} due to invalid hash signature"
+                )
+                .unlessA
+            )
+      }
+
+    private def validateCollateral(batch: RumorBatch): F[RumorBatch] =
+      batch.filterA {
+        case (hash, signedRumor) => {
+          signedRumor.proofs.toNonEmptyList
+            .traverse(proof => collateral.hasCollateral(PeerId.fromId(proof.id)))
+            .map(_.foldLeft(true)(_ && _))
```

Comment:
Instead of iterating over all the proofs you can just find the proof which has no collateral. Do do that you can just:

```scala
signedRumor.proofs.toNonEmptyList
  .findM(proof => collateral.hasCollateral(PeerId.fromId(proof.id)).map(!_))
```

Reply:
Thanks for hint! Instead of `findM` I've used `forallM`.